### PR TITLE
Civic Pass UI integration

### PR DESCRIPTION
**Pull request by dankelleher**
This PR adds UI support for the Civic Pass spl-governance plugin. 
It adds a UI component for obtaining the Civic Pass, as well as transactions for configuring a DAO to use the plugin.

Coming soon, in a later PR, once changes are made in spl-governance to support it: Composable plugins

### CODE REVIEW

File: GatewayPlugin/sdk/accounts.tsx [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  realm: ProgramAccount<Realm>,
+  walletPk: PublicKey
+): Promise<GatewayTokenContext> => {
+  // Get the registrar for the realm
+  const { registrar } = await getGatewayRegistrarPDA(
+    realm.pubkey,
+    realm.account.communityMint,
+    client.program.programId
+  )
+  console.log('GATEWAY: getGatewayToken registrar', registrar)
```
Comment (abrzezinski94) [inline GatewayPlugin/sdk/accounts.tsx:90]: nit: remove console.logs

File: GatewayPlugin/sdk/accounts.tsx [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+  console.log('GATEWAY: getGatewayToken program', client.program.account)
```
Comment (abrzezinski94) [inline GatewayPlugin/sdk/accounts.tsx:91]: nit: remove console.logs

File: GatewayPlugin/sdk/accounts.tsx [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  // @ts-ignore
+  const registrarObject = await client.program.account.registrar.fetch(
```
Comment (abrzezinski94) [inline GatewayPlugin/sdk/accounts.tsx:93]: ts-ignore should be unnecessary why its needed in here ?

File: GatewayPlugin/sdk/accounts.tsx [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    registrar
+  )
+
+  // Find the gatekeeper network from the registrar
+  console.log('GATEWAY: getGatewayToken registrarObject', registrarObject)
```
Comment (abrzezinski94) [inline GatewayPlugin/sdk/accounts.tsx:98]: nit: console logs

File: GatewayPlugin/store/gatewayPluginStore.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    set((s) => {
+      s.state.isLoadingGatewayToken = val
+    })
+  },
+  setGatewayToken: (gatewayToken, votingClient) => {
+    votingClient._setCurrentVoterGatewayToken(gatewayToken)
+    set((s) => {
+      s.state.gatewayToken = gatewayToken
+    })
+    console.log('GATEWAY: setVotingPower')
```
Comment (abrzezinski94) [inline GatewayPlugin/store/gatewayPluginStore.ts:46]: nit: console.log

File: VoteStakeRegistry/sdk/api.ts [hunk 6]
```diff
@@ -39,6 +42,21 @@ export const tryGetNftRegistrar = async (
   }
 }
 
+// TODO merge with the above, perhaps by creating a parameterised client
+export const tryGetGatewayRegistrar = async (
```
Comment (abrzezinski94) [inline VoteStakeRegistry/sdk/api.ts:46]: Shouldn’t it be in gateaway client sdk folder ?

File: actions/castVote.ts [hunk 7]
```diff
@@ -52,6 +52,9 @@ export async function castVote(
     proposal,
     tokeOwnerRecord
   )
+
+  console.log('castVote plugin', plugin)
```
Comment (abrzezinski94) [inline actions/castVote.ts:56]: nit: console log

File: components/TokenBalance/TokenBalanceCardWrapper.tsx [hunk 8]
```diff
@@ -31,6 +36,16 @@ const TokenBalanceCardWrapper = ({
       currentPluginPk && vsrPluginsPks.includes(currentPluginPk?.toBase58())
     const isNftMode =
       currentPluginPk && nftPluginsPks.includes(currentPluginPk?.toBase58())
+    const isGatewayMode =
+      currentPluginPk && gatewayPluginsPks.includes(currentPluginPk?.toBase58())
+
+    console.log('***Is Gateway Mode', {
```
Comment (abrzezinski94) [inline components/TokenBalance/TokenBalanceCardWrapper.tsx:42]: nit: console log

File: hooks/useQueryContext.tsx [hunk 9]
```diff
@@ -5,6 +5,8 @@ export default function useQueryContext() {
   const router = useRouter()
   const { cluster } = router.query
 
+  console.log('router query', router.query)
```
Comment (abrzezinski94) [inline hooks/useQueryContext.tsx:8]: nit: console log

File: hooks/useVotingPlugins.ts [hunk 10]
```diff
@@ -92,6 +107,40 @@ export function useVotingPlugins() {
     }
     setIsLoadingNfts(false)
   }
+  const handleGetGatewayToken = async () => {
+    console.log('GATEWAY: handleGetGatewayToken')
```
Comment (abrzezinski94) [inline hooks/useVotingPlugins.ts:111]: nit: console logs

File: pages/dao/[symbol]/proposal/new.tsx [hunk 11]
```diff
@@ -140,6 +142,7 @@ const New = () => {
   }
 
   const getAvailableInstructionsForIndex = (index) => {
+    console.log('Available instructions', availableInstructions)
```
Comment (abrzezinski94) [inline pages/dao/[symbol]/proposal/new.tsx:145]: nit: console log

File: GatewayPlugin/sdk/accounts.tsx [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    [
+      Buffer.from('voter-weight-record'),
+      realmPk.toBuffer(),
+      mint.toBuffer(),
+      walletPk.toBuffer(),
+    ],
+    clientProgramId
+  )
+
+  console.log('getGatewayVoterWeightRecord', [
```
Comment (abrzezinski94) [inline GatewayPlugin/sdk/accounts.tsx:63]: nit: console log

File: components/Gateway/GatewayCard.tsx [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+  // Temporary, until uniquenessPass is live
+  const gatewayStage = useMemo(() => {
+    const uniquenessPass = 'tunQheuPpHhjjsbrUDp4rikqYez9UXv4SXLRHf9Kzsv'
+    if (gatekeeperNetwork && gatekeeperNetwork.toBase58() === uniquenessPass) {
+      return 'preprod'
+    }
+    return 'prod'
+  }, [gatekeeperNetwork])
+
+  console.log('GatewayCard', {
```
Comment (abrzezinski94) [inline components/Gateway/GatewayCard.tsx:109]: nit console log

File: hooks/useVotingPlugins.ts [hunk 14]
```diff
@@ -163,6 +213,26 @@ export function useVotingPlugins() {
       }
     }
 
+    const handleGatewayPlugin = () => {
+      console.log('GATEWAY: handleGatewayPlugin')
```
Comment (abrzezinski94) [inline hooks/useVotingPlugins.ts:217]: nit console logs

File: hooks/useVotingPlugins.ts [hunk 15]
```diff
@@ -178,19 +248,28 @@ export function useVotingPlugins() {
         }
       }
     }
+
+    console.log('handling plugins...', {
```
Comment (abrzezinski94) [inline hooks/useVotingPlugins.ts:252]: nit console logs

File: GatewayPlugin/sdk/accounts.tsx (hunk 3 above)
Comment (dankelleher) [reply GatewayPlugin/sdk/accounts.tsx:93]: Anchor made my account names lowercase but in the IDL they are still upper case. The code works but ts complains. - I will see if I can manually fix it in the plugin client and remove the ts-ignore

Comment (dankelleher) [reply GatewayPlugin/sdk/accounts.tsx:78]: Fixed

File: GatewayPlugin/sdk/accounts.tsx (hunk 2 above)
Comment (dankelleher) [reply GatewayPlugin/sdk/accounts.tsx]: ✅

File: GatewayPlugin/sdk/accounts.tsx (hunk 1 above)
Comment (dankelleher) [reply GatewayPlugin/sdk/accounts.tsx]: ✅

File: GatewayPlugin/sdk/accounts.tsx (hunk 4 above)
Comment (dankelleher) [reply GatewayPlugin/sdk/accounts.tsx]: ✅

File: GatewayPlugin/store/gatewayPluginStore.ts (hunk 5 above)
Comment (dankelleher) [reply GatewayPlugin/store/gatewayPluginStore.ts]: ✅

File: pages/dao/[symbol]/proposal/new.tsx (hunk 11 above)
Comment (dankelleher) [reply pages/dao/[symbol]/proposal/new.tsx]: ✅

File: GatewayPlugin/sdk/accounts.tsx (hunk 12 above)
Comment (dankelleher) [reply GatewayPlugin/sdk/accounts.tsx]: ✅

File: components/Gateway/GatewayCard.tsx (hunk 13 above)
Comment (dankelleher) [reply components/Gateway/GatewayCard.tsx]: ✅

File: hooks/useVotingPlugins.ts (hunk 14 above)
Comment (dankelleher) [reply hooks/useVotingPlugins.ts]: ✅

File: hooks/useVotingPlugins.ts (hunk 15 above)
Comment (dankelleher) [reply hooks/useVotingPlugins.ts]: ✅

**Comment by abrzezinski94** [comment]
looks good please resolve merge conflicts.

**Pull request merged**